src/patterns.rs

```diff
@@ -99,6 +100,10 @@ pub struct Pattern {
     /// Position of this form among *forms only* in the original (display) order.
     /// This is stable and survives reordering/cloning.
     pub original_index: usize,
+    /// Determine whether the string is deterministic. Created on init.
+    deterministic: bool,
+    /// The set of variables present in the pattern
+    _variables: HashSet<char>,
```

Comment:
why the underscore (here and elsewhere)?

Reply:
I didn't like having both a `.variables` and a `.variables()` but I can be convinced to change it

Reply:
I'll try to come up with something in my next PR...